src/runtime_src/core/pcie/driver/linux/xocl/subdev/xmc.c

```diff
@@ -1307,13 +1311,29 @@ static int get_temp_by_m_tag(struct xocl_xmc *xmc, char *m_tag)
 /* Runtime clock scaling sysfs node */
 static bool scaling_condition_check(struct xocl_xmc *xmc, struct device *dev)
 {
+	if (xmc->sc_presence) {
+		//Feature present bit may configured each time an xclbin is downloaded,
+		//or following a reset of the CMC Subsystem. So, check for latest
+		//status every time.
```

Comment:
why clock scaling feature may change from each time an xclbin is downloaded? why would it change is CMC is reset?

Reply:
Hi Max, there is a change cmc_access_ops() in xmc which was added by David (@xdavidz ) to enable new features when required during xclbin download actully. It is a function callback for xocl_cmc_freeze() in __icap_download_bitstream_axlf() from icap.c.

Reply:
xocl_cmc_freeze (aka cmc_access_ops(.., 0) will only proceed when the IORES_CMC_MUTEX is present after a new xclbin download. I think only u50 blp 1rp has this resource.  What is the original issue that you are fixing?

Reply:
Hi @xdavidz ,
I am adding changes to read the current status of feature present & enable from 0xB20 reg as part of the http://jira.xilinx.com/browse/CR-1051402

Reply:
> 
> 
> Hi @xdavidz ,
> I am adding changes to read the current status of feature present & enable from 0xB20 reg as part of the http://jira.xilinx.com/browse/CR-1051402

I'm not sure if the clock throttling feature can be disabled or enabled by a randomly downloaded xclbin or something else could go wrong? Please work with @xdavidz to resolve this before I can merge this pull request. I've marked @xdavidz as a code reviewer as well.

Reply:
@maxzhen I think the logic here is that after download each NEW xclbin, we set the feature present to 1, then the xbmgmt config --device --cs_threshold_power_override 30 will work, otherwise it won't take effect.

I have verified that after download xclbin, we are able to:
./xbmgmt config --device --cs_threshold_power_override 30
./xbmgmt config --show|grep scaling_thres
        scaling_threshold_power_override: 30W

./xbmgmt config --device --cs_threshold_power_override 50
[root@xsjhemn41 davidz]# ./xbmgmt config --show|grep scaling_thres
        scaling_threshold_power_override: 50W

re download xclbin by xbutil validate -q, and make sure the value has not been changed.

./xbmgmt config --show|grep scaling_thres
        scaling_threshold_power_override: 50W

So it looks ok for me. The only nits:
we should warn user if there is not xclbin (aka now feature present, config won't take effect).

Hence, approve from me.

Reply:
We should allow admin to configure clock throttling when there is no xclbin on the board. @rajkumar-xilinx , please file a CR for this issue and fix it later. For now, we should at least print out a warning message if admin tries to set clock throttling and there is no xclbin on the board. Please make these changes in follow-up pull request. I'll merge this pull request for now.